File: ios/FolderUtils.m
#import "FolderUtils.h"

@implementation FolderUtils {}

+ (NSURL *)createOrRetrieveDirectory:(NSURL *)directoryURL {
    NSFileManager *fileManager = [NSFileManager defaultManager];
    BOOL isDirectory;
    BOOL exists = [fileManager fileExistsAtPath:directoryURL.path isDirectory:&isDirectory];
    
    if (!exists || !isDirectory) {
        NSError *error;
        BOOL created = [fileManager createDirectoryAtURL:directoryURL
                             withIntermediateDirectories:YES
                                              attributes:nil
                                                   error:&error];
        if (!created) {
            NSLog(@"FolderUtils: Failed to create directory at %@. Error: %@", directoryURL.path, error);
            return nil;
        }
    }
    
    return directoryURL;
}

+ (NSString *)documentsDirectory {
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    return (paths.count) ? paths[0] : nil;
}

+ (NSURL *)otaDirectory {
    NSURL *otaDirectory = [[NSURL fileURLWithPath:[self documentsDirectory]] URLByAppendingPathComponent:@"ota"];
    NSURL *createdDirectory = [self createOrRetrieveDirectory:otaDirectory];
    return createdDirectory;
}

+ (NSURL *)downloadDestinationDirectory {
    NSURL *downloadDestinationDirectory = [[FolderUtils otaDirectory] URLByAppendingPathComponent:@"download"];
    NSURL *createdDirectory = [self createOrRetrieveDirectory:downloadDestinationDirectory];
    return createdDirectory;
}

+ (NSURL *)destinationDirectoryForVersion:(NSString *)version {
    NSURL *versionDirectory = [[self otaDirectory] URLByAppendingPathComponent:version];
    return [self createOrRetrieveDirectory:versionDirectory];
}

+ (NSURL *)destinationDirectoryForVersion:(NSString *)version fileName:(NSString *)fileName {
    NSURL *versionDirectory = [self destinationDirectoryForVersion:version];
    if (!versionDirectory) {
        NSLog(@"FolderUtils: Failed to get or create version directory for %@", version);
        return nil;
    }
    
    NSURL *fileNameDirectory = [versionDirectory URLByAppendingPathComponent:fileName];
    return [self createOrRetrieveDirectory:fileNameDirectory];
}

+ (NSArray<NSString *> *)getAllFilesInDirectory:(NSURL *)directory {
    NSError *error;
    NSArray<NSString *> *files = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:directory.path error:&error];
    if (error) {
        NSLog(@"FolderUtils: Error getting contents of directory %@: %@", directory, error);
        return @[];
    }
    return files;
}

@end

File: android/src/main/java/com/ota/FolderUtils.java
package com.ota;

import android.content.Context;
import java.io.File;
import android.util.Log;

public class FolderUtils {
    private static final String TAG = "FolderUtils";
    private static final String OTA_DIR = "ota";
    private static final String DOWNLOAD_DIR = "download";

    public static File getOtaDirectory(Context context) {
        File otaDir = new File(context.getFilesDir(), OTA_DIR);
        if (!otaDir.exists()) {
            if (!otaDir.mkdirs()) {
                Log.e(TAG, "Failed to create OTA directory");
                return null;
            }
        }
        return otaDir;
    }

    public static File getDownloadDirectory(Context context) {
        File downloadDir = new File(getOtaDirectory(context), DOWNLOAD_DIR);
        if (!downloadDir.exists()) {
            if (!downloadDir.mkdirs()) {
                Log.e(TAG, "Failed to create download directory");
                return null;
            }
        }
        return downloadDir;
    }

    public static File getVersionDirectory(Context context, String version) {
        File versionDir = new File(getOtaDirectory(context), version);
        if (!versionDir.exists()) {
            if (!versionDir.mkdirs()) {
                Log.e(TAG, "Failed to create version directory");
                return null;
            }
        }
        return versionDir;
    }

    public static File[] getAllFilesInDirectory(File directory) {
        if (directory == null || !directory.exists()) {
            return new File[0];
        }
        File[] files = directory.listFiles();
        return files != null ? files : new File[0];
    }
} 